addons/website/models/mixins.py

```diff
@@ -191,35 +197,220 @@ def _compute_website_url(self):
             record.website_url = '#'
 
+    @api.constrains("is_published")
+    def _check_is_published(self):
+        for record in self:
+            if record.is_published and not record.published_date:
+                # Check if something must be done just after publishing.
+                self._check_for_action_post_publish()
+                # If the record is being published and does not have a
+                # published_date set, set the published_date to the current date
+                # and time.
+                record.published_date = fields.datetime.now()
+                # Additionally, set the publish_on field to False
+                # This likely means that the record should be published
+                # immediately.
+                record.publish_on = False
+
     def _default_is_published(self):
         return False
 
+    def action_unschedule(self):
+        for page in self:
+            page.publish_on = False
+
+    def _models_generator(self):
+        """
+        This method generates a sequence of models that contain the field
+        'publish_on'. Thus all the models that inherit WebsitePublishedMixin.
+
+        :return: A generator yielding models containing the 'publish_on' field.
+        :rtype: generator of odoo.models.Model
+        """
+        # Retrieve the models containing the field 'publish_on' and filter
+        # transient models and related fields.
+        models = [
+            self.env[m]
+            for m in self.env["ir.model.fields"]
+            .sudo()
+            .search(
+                [
+                    ("name", "=", "publish_on"),
+                    ("model_id.transient", "=", False),
+                    ("related", "=", False),
+                ]
+            )
+            .mapped("model")
+            if (
+                m in self.env
+                and {"id", "is_published"} <= set(self.env[m]._fields)
+            )
+        ]
+        yield from models
+
+    def _cron_publish_scheduled_pages(self):
+        """
+        Method triggered by a cron job to publish records at a scheduled time.
+
+        This method iterates through each model containing the 'publish_on'
+        field. It searches for records scheduled to be published before the
+        current datetime. If such records are found, it updates their
+        'is_published' field to True and clears the 'publish_on' field.
+
+        :return: None
+        :rtype: None
+        """
+        # Iterate through each model containing the 'publish_on' field.
+        for model in self._models_generator():
+            pages = model.search(
+                [
+                    ("publish_on", "!=", False),
+                    ("publish_on", "<", fields.datetime.now()),
+                ]
+            )
+            # Update 'is_published' field to True for retrieved records.
+            if pages:
+                pages.write({"is_published": True, "publish_on": False})
```

Comment:
This could also reset "published_date" to False so that the new publish date is written in `_check_is_published`. I think you generally want the date of the actual publication, not an old date that might be there for several reasons. What do you think?